[CodeCompletion] Don't provide 'init(nilLiteral:)' et al in optional context

rintaro opened the pull request:
For unresolved member completion, member decls in `Optional` referenceable with implicit member expression are useless in most cases.
- `init(_: <#Wrapped#>)`
- `init(nilLiteral: ())`
Instead, provide `nil` keyword with erasing `.` instruction.

As for `.some(<#Wrapped#>)` and `.none`, although they aren't useful in expression context, it's common to use them in pattern context. So don't hide them until we implement a mechanism to distinguish them.

rdar://problem/47806831

rintaro committed:
[CodeCompletion] Don't provide 'init(nilLiteral:)' et al in optional context

For unresolved member completion, member decls in `Optional`
referenceable by implicit member expression are useless in most cases.
- `init(_: <#Wrapped#>)`
- `init(nilLiteral: ())`
- `.some(<#Wrapped#>)`
- `.none`

Instead, provide `nil` with erasing `.` instruction.

rdar://problem/47806831

jrose-apple commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -3634,6 +3637,14 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
+      // '.some(<some>)' and '.none'. They are useless in most cases.
+      if (T->getOptionalObjectType() &&
+          VD->getModuleContext()->isStdlibModule() &&
+          (isa<EnumElementDecl>(VD) || isa<ConstructorDecl>(VD))) {
```
Ah, you probably don't want to ignore *all* constructors. Someone could have added their own.

rintaro replied:
I think checking `VD->getModuleContext()->isStdlibModule()` is enough to limit them to `init(nilLiteral: ())` and `init(<#Wrapped#>)`. Added test case for that.
Is there any cases which hide user own initializers?

jrose-apple replied:
That should be good!

jrose-apple replied:
OTOH, maybe we should "just" have a `hideFromCodeCompletion` attribute…

benlangmuir commented:
> init(_: <#Wrapped#>)
> init(nilLiteral: ())

Okay

> .some(<#Wrapped#>)
> .none

What's the justification for hiding these?  I've seen code that uses this style, particularly in switch cases.  It's more discoverable than "?" in a pattern binding.  Users from other languages with Optional may also be used to writing things this way.

rintaro commented:
Hm, I was thinking mostly about expression context. But yes, `.some()`/`.none` is commonly used in matching pattern and I agree we should not hide them in pattern context.
I think, eventually, we should distinguish between pattern and expression context.

rintaro committed:
[CodeCompletion] Don't ignore .some(Wrapped) and .none for now

with "TODO: ignore them in expression context".
They are useful in pattern context, so we should provide them in
completion.

rintaro commented:
Restored `.some` and `.none` for now with `TODO` note.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test
